drivers/SPI.cpp

```diff
@@ -23,33 +24,107 @@
 
 #if DEVICE_SPI
 
+/* Backwards compatibility with HALs that don't provide this */
+MBED_WEAK SPIName spi_get_peripheral_name(PinName /*mosi*/, PinName /*miso*/, PinName /*mclk*/)
+{
+    return (SPIName)1;
+}
+
 namespace mbed {
 
-#if DEVICE_SPI_ASYNCH && TRANSACTION_QUEUE_SIZE_SPI
-CircularBuffer<Transaction<SPI>, TRANSACTION_QUEUE_SIZE_SPI> SPI::_transaction_buffer;
-#endif
+SPI *SPI::_list;
 
 SPI::SPI(PinName mosi, PinName miso, PinName sclk, PinName ssel) :
-    _spi(),
 #if DEVICE_SPI_ASYNCH
     _irq(this),
-    _usage(DMA_USAGE_NEVER),
-    _deep_sleep_locked(false),
 #endif
-    _bits(8),
-    _mode(0),
-    _hz(1000000),
-    _write_fill(SPI_FILL_CHAR)
+    _mosi(mosi),
+    _miso(miso),
+    _sclk(sclk),
+    _hw_ssel(ssel),
+    _sw_ssel(NC)
+{
+    _do_construct();
+}
+
+SPI::SPI(PinName mosi, PinName miso, PinName sclk, PinName ssel, use_gpio_ssel_t) :
+#if DEVICE_SPI_ASYNCH
+    _irq(this),
+#endif
+    _mosi(mosi),
+    _miso(miso),
+    _sclk(sclk),
+    _hw_ssel(NC),
+    _sw_ssel(ssel, 1)
+{
+    _do_construct();
+}
+
+void SPI::_do_construct()
 {
     // No lock needed in the constructor
-    spi_init(&_spi, mosi, miso, sclk, ssel);
+#if DEVICE_SPI_ASYNCH
+    _usage = DMA_USAGE_NEVER;
+    _deep_sleep_locked = false;
+#endif
+    _select_count = 0;
+    _bits = 8;
+    _mode = 0;
+    _hz = 1000000;
+    _write_fill = SPI_FILL_CHAR;
+
+    SPIName name = spi_get_peripheral_name(_mosi, _miso, _sclk);
+
+    // Searching for and creation of control blocks must be synchronised
+    singleton_lock();
+    // Find an existing user of this peripheral, and share their control block
+    _peripheral = SPI::_lookup_peripheral(name);
+    // Otherwise, make a new one
+    if (!_peripheral) {
+        _peripheral.reset(new Peripheral(name));
```

Comment:
The extra code size from using dynamic memory here will likely be a problem for bootloaders.